src/piton/featurizers/core.py

```diff
@@ -0,0 +1,420 @@
+    ), f"`data` should have equal shape as `indices`, but instead have {np_data.shape} != {np_indices.shape}"
+    data_matrix = scipy.sparse.csr_matrix(
+        (np_data, np_indices, np_indptr), shape=(total_rows, total_columns)
+    )
+
+    label_pids: NDArray[Literal["n_labels, 1"], np.int64] = np.array(
+        [x[0] for x in label_data], dtype=np.int64
+    )
+    label_values: NDArray[Literal["n_labels, 1"], Any] = np.array(
+        [x[1] for x in label_data]
+    )
+    label_times: NDArray[Literal["n_labels, 1"], np.datetime64] = np.array(
+        [x[2] for x in label_data], dtype=np.datetime64
+    )
+    assert (
+        label_pids.shape == label_values.shape == label_times.shape
+    ), f"These should all be equal: {label_pids.shape} | {label_values.shape} | {label_times.shape}"
+
+    # data_matrix.check_format() # remove when we think its works
+
+    return data_matrix, label_pids, label_values, label_times
+
+
+def _run_preprocess_featurizers(
+    args: Tuple[str, List[int], LabeledPatients, List[Featurizer]]
+) -> List[Featurizer]:
+    """Apply preprocessing of featurizers to the set of patients included in `patient_ids`.
+    Gets called as a parallelized subprocess of the .preprocess_featurizers() method of `FeaturizerList`.
+    """
+    database_path: str = args[0]
+    patient_ids: List[int] = args[1]
+    labeled_patients: LabeledPatients = args[2]
+    featurizers: List[Featurizer] = args[3]
+
+    # Load patients
+    database: PatientDatabase = PatientDatabase(database_path)
+
+    # Preprocess featurizers on all Labels for each Patient...
+    for patient_id in patient_ids:
+        patient: Patient = database[patient_id]  # type: ignore
+        labels: List[Label] = labeled_patients.get_labels_from_patient_idx(
+            patient_id
+        )
+
+        if len(labels) == 0:
+            continue
+
+        # Preprocess featurizers
+        for featurizer in featurizers:
+            if featurizer.is_needs_preprocessing():
+                featurizer.preprocess(patient, labels)
+
+    return featurizers
+
+
+class Featurizer(ABC):
+    """A Featurizer takes a Patient and a list of Labels, then returns a row for each timepoint.
+    Featurizers must be preprocessed before they are used to compute normalization statistics.
+    A sparse representation named ColumnValue is used to represent the values returned by a Featurizer.
+    """
+
+    # TODO - rename to 'train' ??
+    def preprocess(self, patient: Patient, labels: List[Label]):
+        """Preprocess the featurizer on the given patient and label indices.
+        This should do nothing if `is_needs_preprocessing()` returns FALSE,
+        i.e. the featurizer doesn't need preprocessing.
+
+        Args:
+            patient (Patient): A patient to preprocess on.
+            labels (List[Label]): The list of labels of this patient to preprocess on.
+        """
+        pass
+
+    @classmethod
+    def aggregate_featurizers(
```

Comment:
I don't think a default implementation makes sense for this function. Should be abstract.